image_tf_encoder.py

```diff
@@ -0,0 +1,118 @@
+__copyright__ = "Copyright (c) 2020-2021 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from typing import Union, Iterable, List, Any, Optional, Dict
+
+import numpy as np
+from jina import DocumentArray, Executor, requests
+from jina.logging.logger import JinaLogger
+
+
+def _batch_generator(data: List[Any], batch_size: Optional[int]):
+    for i in range(0, len(data), batch_size):
+        yield data[i:i + batch_size]
+
+
+class ImageTFEncoder(Executor):
+    """
+    :class:`ImageKerasEncoder` encodes ``Document`` content from a ndarray,
+    potentially B x (Channel x Height x Width) into a ndarray of `B x D`.
+
+    Where `B` is the batch size and `D` is the Dimension.
+
+    The :class:`ImageKerasEncoder` wraps the models from
+    `tensorflow.keras.applications`. <https://keras.io/applications/>`_.
+
+    :param model_name: the name of the model. Supported models include
+        ``DenseNet121``, ``DenseNet169``, ``DenseNet201``,
+        ``InceptionResNetV2``, ``InceptionV3``, ``MobileNet``,
+        ``MobileNetV2``, ``NASNetLarge``, ``NASNetMobile``,
+        ``ResNet101``, ``ResNet152``, ``ResNet50``, ``ResNet101V2``,
+        ``ResNet152V2``, ``ResNet50V2``, ``VGG16``, ``VGG19``,
+        ``Xception``,
+    :param img_shape: The shape of the image to be encoded.
+    :param pool_strategy: the pooling strategy. Options are:
+        - `None`: Means that the output of the model will be the 4D tensor
+            output of the last convolutional block.
+        - `avg`: ;eans that global average pooling will be applied to the
+            output of the last convolutional block, and thus the output of
+            the model will be a 2D tensor.
+        - `max`: Means that global max pooling will be applied.
+    :param default_batch_size: size of each batch
+    :param default_traversal_paths: traversal path of the Documents, (e.g. 'r', 'c')
+    :param on_gpu: set to True if using GPU
+    :param args: additional positional arguments.
+    :param kwargs: additional positional arguments.
+    """
+
+    def __init__(self,
+                 model_name: str = 'MobileNetV2',
+                 img_shape: int = 336,
+                 pool_strategy: str = 'max',
+                 default_batch_size: int = 32,
+                 default_traversal_paths: List[str] = ['r'],
```

Comment:
Don't use list as default in Python, this can lead to serious bugs which are very weird.
See [here](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) for more details.
Just use `None` as default and set ['r'] inside the function if default_traversal_paths is None

Reply:
good point